File: 백준/Gold/1082. 방 번호/방 번호.py
import sys
input = sys.stdin.readline

n = int(input())
price = list(map(int, input().split()))
m = int(input())

dp = [0] * (m+1)
for num in range(n-1, -1, -1):
    for money in range(price[num], m+1):
        dp[money] = max(dp[money], dp[money-price[num]]*10 + num, num)

print(dp[-1])

File: 백준/Gold/1082. 방 번호/방 번호.js
const input = require('fs').readFileSync(process.platform === 'linux' ? '/dev/stdin' : __dirname + '/input.txt').toString().split('\n')

const N = BigInt(input[0])
const P = input[1].split(' ').map((n) => BigInt(n))
const M = Number(input[2])

const dp = Array(M + 1).fill(0n)
for (let num = N; num >= 0n; num--) {
  for (let m = P[num]; m <= M; m++) {
    const money = dp[m - P[num]] * 10n + num
    if (dp[m] < money) dp[m] = money
    
  }
}
console.log(dp[M].toString())